src/gui/vectortile/qgsvectortilebasiclabelingwidget.cpp

```diff
@@ -0,0 +1,458 @@
+
+    void insertStyle( int row, const QgsVectorTileBasicLabelingStyle &style );
+
+    // drag'n'drop support
+    Qt::DropActions supportedDropActions() const override;
+    QStringList mimeTypes() const override;
+    QMimeData *mimeData( const QModelIndexList &indexes ) const override;
+    bool dropMimeData( const QMimeData *data, Qt::DropAction action, int row, int column, const QModelIndex &parent ) override;
+
+  private:
+    QgsVectorTileBasicLabeling *mLabeling = nullptr;
+};
+
+
+QgsVectorTileBasicLabelingListModel::QgsVectorTileBasicLabelingListModel( QgsVectorTileBasicLabeling *l, QObject *parent )
+  : QAbstractListModel( parent )
+  , mLabeling( l )
+{
+}
+
+int QgsVectorTileBasicLabelingListModel::rowCount( const QModelIndex &parent ) const
+{
+  if ( parent.isValid() )
+    return 0;
+
+  return mLabeling->styles().count();
+}
+
+int QgsVectorTileBasicLabelingListModel::columnCount( const QModelIndex & ) const
+{
+  return 5;
+}
+
+QVariant QgsVectorTileBasicLabelingListModel::data( const QModelIndex &index, int role ) const
+{
+  if ( index.row() < 0 || index.row() >= mLabeling->styles().count() )
+    return QVariant();
+
+  const QList<QgsVectorTileBasicLabelingStyle> styles = mLabeling->styles();
+  const QgsVectorTileBasicLabelingStyle &style = styles[index.row()];
+
+  switch ( role )
+  {
+    case Qt::DisplayRole:
+    {
+      if ( index.column() == 0 )
+        return style.styleName();
+      else if ( index.column() == 1 )
+        return style.layerName().isEmpty() ? tr( "(all layers)" ) : style.layerName();
+      else if ( index.column() == 2 )
+        return style.minZoomLevel() >= 0 ? style.minZoomLevel() : QVariant();
+      else if ( index.column() == 3 )
+        return style.maxZoomLevel() >= 0 ? style.maxZoomLevel() : QVariant();
+      else if ( index.column() == 4 )
+        return style.filterExpression().isEmpty() ? tr( "(no filter)" ) : style.filterExpression();
+
+      break;
+    }
+
+    case Qt::EditRole:
+    {
+      if ( index.column() == 0 )
+        return style.styleName();
+      else if ( index.column() == 1 )
+        return style.layerName();
+      else if ( index.column() == 2 )
+        return style.minZoomLevel();
+      else if ( index.column() == 3 )
+        return style.maxZoomLevel();
+      else if ( index.column() == 4 )
+        return style.filterExpression();
+
+      break;
+    }
+
+    case Qt::CheckStateRole:
+    {
+      if ( index.column() != 0 )
+        return QVariant();
+      return style.isEnabled() ? Qt::Checked : Qt::Unchecked;
+    }
+
+  }
+  return QVariant();
+}
+
+QVariant QgsVectorTileBasicLabelingListModel::headerData( int section, Qt::Orientation orientation, int role ) const
+{
+  if ( orientation == Qt::Horizontal && role == Qt::DisplayRole && section >= 0 && section < 5 )
+  {
+    QStringList lst;
+    lst << tr( "Label" ) << tr( "Layer" ) << tr( "Min. zoom" ) << tr( "Max. zoom" ) << tr( "Filter" );
```

Comment:
What's our convention re capitalization in headers? I'm guessing it's Title Case, but not 100% sure....

Reply:
According to point 3 in https://docs.qgis.org/3.10/en/docs/developers_guide/hig.html it should be Title Case, but I just copied the convention from what is already being used in rule-based renderer/labeling widget. If you prefer Title Case then I guess I should also update the existing texts there (which will break translations)